### PR TITLE
Add `--secret` flag to `cog build` and `cog push` commands

### LINKED ISSUE #1009: Build secrets
This syntax:

```
build:
  run:
    - command: pip install ...
      mounts:
       - type: secret
         id: someid
         target: path/in/container
```

Should turn into:

```
RUN --mount=type=secret,id=someid,target=path/in/container pip install ...
```

then to build you would use:

```
cog build --secret id=someid,src=path/in/host
```

todo:

- [ ] end-to-end documentation on this feature
- [ ] enable feature to work without swarm
- [ ] ensure `cog build` re-builds when contents of secret file changes

### PR DESCRIPTION
**Pull request by mattt**
Resolves #1009

### CODE REVIEW

**Comment by anotherjesse** [comment]
@mattt the code lgtm

I think we should add to the docs.yaml how to use it:

```
cog build --secret id=foo,source=/path/to/foo
```

I want to add a note about how caching secrets works since it is different from files, but I'm not sure the best recommendation for telling users how to do so 

`docker rmi cog-{directory-name}` or `cog rmi r8.im/username/model_name` only deletes the top layer, but the underlying cached layers still remain, and the build still has the old secret.

From googling/chatgpting it seems like best practice is to use ARG before your RUN

```
ARG SECRET_VERSION
RUN --mount-type=secret ....
```

The other way is `--no-cache`, which as Ben mentioned we would need to add support for.

I'm only hesitant to ship without documentation/process for breaking cache as I ran into the issue within minutes of using it.

**Comment by mattt** [comment]
@anotherjesse Agreed on the point about docs. I'll add that now.

I've opened PRs to this PR branch implementing both approaches you described:

- https://github.com/replicate/cog/pull/1014
- https://github.com/replicate/cog/pull/1015

File: docs/private-package-registry.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Private package registry
+
+This guide describes how to build a Docker image with Cog that fetches Python packages from a private registry during setup.
+
+## `pip.conf`
+
+Create a `pip.conf` file with an `index-url` set to the registry's URL with embedded credentials.
```
Comment (anotherjesse) [inline docs/private-package-registry.md:7]: Great overview!

Question: Should we advise to either have a .dockerignore or (preferably?) keep this file outside of the cog project directory - as all content gets copied in your `cog build` step?

(Many times when I review customers' images they have lots of extra stuff that the they don't realize is copied in automatically.)

File: docs/private-package-registry.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+          id: pip
+          target: /etc/pip.conf
+```
+
+## Build
+
+When building or pushing your model with Cog, pass the `--secret` option with an `id` matching the one specified in `cog.yaml`, along with a path to your local `pip.conf` file.
+
+```console
+$ cog build cog build --secret id=pip,source=/path/to/pip.conf
```
Comment (anotherjesse) [inline docs/private-package-registry.md:33]: extra `cog build` here

File: docs/private-package-registry.md (hunk 1 above)
Comment (mattt) [reply docs/private-package-registry.md]: Great idea. I just made a few additions that call this out explicitly.

**Comment by anotherjesse** [comment]
@mattt this works great.  I tested by using github private packages

https://docs.readthedocs.io/en/stable/guides/private-python-packages.html#from-a-git-repository

thanks for pushing it over the line.

**Review by bfirsh — APPROVED** [review]
Fantastic. Haven't done a detailed code review but sounds good in principle. This maps to the same `--secret` argument in `docker build`, doesn't it?

**Comment by mattt** [comment]
> Fantastic. Haven't done a detailed code review but sounds good in principle. This maps to the same `--secret` argument in `docker build`, doesn't it?

Correct. Here's where we're passing them as args to that command: https://github.com/replicate/cog/pull/1013/files#diff-7579bff2107ab04f17023ce17e83ca76171c266529245b9a263fcf8eba2a5c65R23-R25.

**Pull request merged**